### ISSUE TITLE
Move away from storing data in the background page

### ISSUE THREAD

**Issue by DNin01** [ScratchAddons/ScratchAddons#6991]
#1538 depends on this issue being resolved. Head there for more details and discussion about this problem.

### What's the problem?

Currently, Scratch Addons stores all its configuration data like what addons are enabled and their settings, every addon manifest and more in memory persistently via a persistent background page. However, this has gone against the best practices for Chrome extensions for some time now[^1] (the main reason we're still using one is because the `webRequest` API requires a persistent background page) and won't be feasible after migrating to Manifest V3. Why?

Well, MV3 replaces background pages with service workers, which always automatically terminate after 30 seconds of not doing anything[^2]. When this happens, all of their variables also vanish, therefore, we can't use it as a persistent memory namespace.

> [!IMPORTANT]
> In general, [it is not okay to keep a service worker alive indefinitely](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#keep_a_service_worker_alive_continuously).

### How to fix it

So, the next step is probably to find a way to load the addons that the user has enabled and apply their settings with **only content scripts and Chrome storage**.

One of the things we'll have to change is how we store data that must be generated or initialized by a script before it can be used for other purposes (like running addons) or to speed things up. Currently, the [`scratchAddons` object](https://github.com/ScratchAddons/ScratchAddons/blob/2b08686ecaedccbacc5570d985c477ce773afc8c/background/declare-scratchaddons-object.js#L5), contained within the background page, is what stores all of the "compiled" data about addons in memory.

Without the ability to rely on the background page to store this data, we'd instead save it to storage, probably session storage, for quick access later. We may not even need the background page to implement this, just a function that generates this data that can be imported into content scripts and the Scratch Addons extension webpages, **no background context involved**. If that data has already been generated, simply load the generated data from storage instead. Running procedures only when necessary is good for performance and is [recommended](https://developer.chrome.com/docs/extensions/mv2/performance#defer_everything).

Also worth considering is whether we want this data to persist in the background **at all times**. Currently, Scratch Addons stores all of its important data like user settings and addon manifests in memory _the whole time_ it's running. It's not much (about half a megabyte, and about 25% of the entire background page's memory usage, from my testing with browser developer tools) and I don't know what the Chrome team recommends when it comes to the memory footprint of an extension, but in my opinion, this is still worth thinking about, considering that Chrome is taking steps to free up browser resources by having extensions move to non-persistent service workers. We could clear this data if the user hasn't been on Scratch or opened Scratch Addons for a couple minutes, just to align with the best practices about persistence. It would be regenerated as necessary. But that's up for discussion.

Also, take care when storing authentication-related data like Scratch account X-Tokens.

This theoretically wouldn't require us to migrate to a service worker at the same time, we'd just move away from having the background page store all Scratch Addons' important variables. And of course, we'd still have a background context.

> [!NOTE]
> - The quota for `storage.session` in Chrome used to be 1 MB, but in Chrome 112, it was increased to 10 MB[^3]. Similarly, `storage.local` used to have a quota of 5 MB, but it was increased to 10 MB in Chrome 114.
> - Firefox added support for `storage.session` in version 115[^4]. Like Chrome, it can store up to 10 MB.

### Related

- #4561
- #5190
- #1536

[^1]: https://developer.chrome.com/docs/extensions/mv2/background-migration
[^2]: https://developer.chrome.com/docs/extensions/develop/concepts/service-workers/lifecycle#idle-shutdown
[^3]: https://developer.chrome.com/docs/extensions/reference/api/storage#storage_areas
[^4]: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/session#browser_compatibility

**Comment by WorldLanguages** [comment]
I believe the persistence aspect of service workers won't force us to make any drastic changes when migrating to MV3. But surely it's nice to consider all of this for the future.

Also DNin, please continue to ask questions and dig on all aspects of the extension, as you write spectacular documentation in the form of issues :P

**Comment by DNin01** [comment]
Since `storage.session` is only supported in Firefox 115+, do you think we could replace it with `storage.local` in those versions instead? Getting which storage namespace to use would probably be as simple as:
```js
const tempStorage = browser.storage.session ? browser.storage.session : browser.storage.local
```
Not sure if the fact that it's stored on the disk would become a problem, though. What if you had a slooooow HDD or another program was actively using the disk?

**Comment by DNin01** [comment]
Another thing to consider is how the `addon.settings` API should work. It seems like it might depend on the background page being active.
https://github.com/ScratchAddons/ScratchAddons/blob/b48fbcb5bfbe8ef1499a02d7cffae35c1c6adaba/addon-api/common/Settings.js#L19
That means we'd have to make some changes if we want dynamically-changeable settings.

**Comment by WorldLanguages** [comment]
> It seems like it might depend on the background page being active.

@DNin01 Are you referring to addon userscripts using `addon.settings.get()` in cases where the service worker is currently not active? Fortunately, that will not be a problem. You can even try it out yourself with the console. Within the context of a tab, `scratchAddons.globalState.addonSettings` will continue to exist even if the extension is put to sleep, is disabled, uninstalled, or whatever.
Most (if not all) of the `scratchAddons.globalState` properties have a local copy on each tab - the background context notifies all open tabs if it suffers changes, if not it is assumed to not have changed.
Think about it. If it wasn't for this local copy, it would be impossible to provide an `addon.settings.get()` method that was synchronous - userscripts would need to use "await" or other promise utilities. Since this get() method just retrieves data from a local object (within the same scripting context) there was no need to make it asynchronous.

**Comment by DNin01** [comment]
> `scratchAddons.globalState.addonSettings` will continue to exist even if the extension is put to sleep, is disabled, uninstalled, or whatever.

Right! So the instance of the `scratchAddons` object that userscripts have access to is stored in the page's `window` object or something?

But dynamic changes might be a problem - I have before reloaded Scratch Addons and then had to reload the page for any setting changes or enabled/disabled addons to apply.

**Comment by WorldLanguages** [comment]
> I have before reloaded Scratch Addons and then had to reload the page for any setting changes or enabled/disabled addons to apply.

I believe that's intentional behavior by Chrome - an extension content script can only talk and receive messages from itself. If the extension is reloaded, it's considered to be a different extension.
When developing in Firefox, this is not the case AFAIK.

For Manifest V3, when the service worker is put to sleep and then wakes up again, it's still considered the same extension, so it shouldn't cause any communication issues.
Our background page notifies every tab when there's some global update, such as a setting that changed.

**Comment by DNin01** [comment]
This should probably get its own issue, but one way to reduce background resource consumption is to limit the notification background process. It's probably okay to frequently ping for unread messages while the user is using Scratch, but we could gradually decrease the ping rate as the user becomes less active, to reduce background resource consumption. My idea is:

- Once immediately on browser restart
- While using Scratch, 2 times per minute, as we always have
- After a few minutes of inactivity, once every 2-3 minutes
- After 15 minutes of inactivity or after a browser restart, once every 15 minutes

Or we could make Scratch Notifier a standalone app.

**Comment by BlueGoat11** [comment]
I think that would workOn Mar 2, 2024, at 4:18 PM, DNin01 ***@***.***> wrote:﻿
This should probably get its own issue, but one way to reduce background resource consumption is to limit the notification background process. It's probably okay to frequently ping for unread messages while the user is using Scratch, but we could gradually decrease the ping rate as the user becomes less active, to reduce background resource consumption. My idea is:

Once immediately on browser restart
While using Scratch, 2 times per minute, as we always have
After a few minutes of inactivity, once every 2-3 minutes
After 15 minutes of inactivity or after a browser restart, once every 15 minutes

Or we could make Scratch Notifier a standalone app.

**Comment by Secret-chest** [comment]
15 minutes is too much, maybe 5? Or make it customisable

**Comment by BlueGoat11** [comment]
How bout 7 min

**Comment by DNin01** [comment]
> 15 minutes is too much, maybe 5? Or make it customisable

I know, I don't really like the sound of fetching messages this infrequently, but at the same time, Google doesn't really like background tasks that are active persistently. I guess it's not truly persistent when it's awake for a small potion of time, though, but reloading the service worker often also gets taxing.

5 minute intervals might be good. 🤔

It's not up to me to decide how to handle this, though.

**Comment by BroJac5246** [comment]
> > 15 minutes is too much, maybe 5? Or make it customisable
> 
> I know, I don't really like the sound of fetching messages this infrequently, but at the same time, Google doesn't really like background tasks that are active persistently. I guess it's not truly persistent when it's awake for a small potion of time, though, but reloading the service worker often also gets taxing.
> 
> 5 minute intervals might be good. 🤔
> 
> It's not up to me to decide how to handle this, though.

I'll just chime in and say that I want it to check frequently if I'm having a conversation with someone in the comments, but after that's been done for, say, half an hour, I'm cool with it checking much less frequently, say every 15 minutes.

**Comment by WorldLanguages** [comment]
> Google doesn't really like background tasks that are active persistently

I don't think there's any problem with using `chrome.alarms` to check for new messages every 1 minute...?
It would make sense to check less often if we want to help to reduce the load on Scratch servers. But I haven't seen any evidence of Google Chrome Web Store review or guidelines saying anything about alarms that run once a minute permanently.
